node/pkg/ethereum/moonbeamimpl.go

```diff
@@ -0,0 +1,226 @@
+	if e.BaseEth.client == nil {
+		panic("client is not initialized!")
+	}
+	if e.mbRpcCon == nil {
+		panic("mbRpcCon is not initialized!")
+	}	
+
+	latestBlock, err := e.getBlock(ctx, nil)
+	if err != nil {
+		return nil, err
+	}
+	currentBlockNumber := *latestBlock.Number
+
+	const DELAY_IN_MS = 250
+	var BIG_ONE = big.NewInt(1)
+
+	timer := time.NewTimer(time.Millisecond) // Start immediately.
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				for {
+					var block *common.NewBlock
+					var err error
+
+					// See if the next block has been created yet.
+					if currentBlockNumber.Cmp(latestBlock.Number) > 0 {
+						latestBlock, err = e.getBlock(ctx, nil)
+						if err != nil {
+							e.logger.Error("failed to look up latest block", zap.String("eth_network", e.BaseEth.NetworkName),
+								zap.Uint64("block", currentBlockNumber.Uint64()), zap.Error(err))
+							break
+						}
+
+						if currentBlockNumber.Cmp(latestBlock.Number) > 0 {
+							// We have to wait for this block to become available.
+							break
+						}
+
+						if currentBlockNumber.Cmp(latestBlock.Number) == 0 {
+							block = latestBlock
+						}
+					}
+
+					// Fetch the hash every time, in case it changes due to a rollback. The only exception is if we just got it above.
+					if block == nil {
+						block, err = e.getBlock(ctx, &currentBlockNumber)
+						if err != nil {
+							e.logger.Error("failed to get current block", zap.String("eth_network", e.BaseEth.NetworkName),
+								zap.Uint64("block", currentBlockNumber.Uint64()), zap.Error(err))
+							break
+						}
+					}
+
+					finalized, err := e.isBlockFinalized(ctx, block.Hash.Hex())
+					if err != nil {
+						e.logger.Error("failed to see if block is finalized", zap.String("eth_network", e.BaseEth.NetworkName),
+							zap.Uint64("block", currentBlockNumber.Uint64()), zap.Error(err))
+						break
+					}
+					if !finalized {
+						break
+					}
+
+					sink <- block
+					currentBlockNumber.Add(&currentBlockNumber, BIG_ONE)
+				}
+
+				timer = time.NewTimer(DELAY_IN_MS * time.Millisecond)
+			}
+		}
+	}()
+
+	sub := &MoonbeamSubscription{
+		err: make(chan error, 1),
+	}
+
+	return sub, err
+}
+
+func (e *MoonbeamImpl) getBlock(ctx context.Context, number *big.Int) (*common.NewBlock, error) {
+	var numStr string
+	if number != nil {
+		numStr = ethHexUtils.EncodeBig(number)
+	} else {
+		numStr = "latest"
+	}
+
+	type Marshaller struct {
+		Number *ethHexUtils.Big
+		Hash ethCommon.Hash `json:"hash"`
+	}
+
+	var m Marshaller
+	err := e.mbRpcCon.CallContext(ctx, &m, "eth_getBlockByNumber", numStr, false)
+	if err != nil {
+		e.logger.Error("failed to get block", zap.String("eth_network", e.BaseEth.NetworkName),
+			zap.String("requested_block", numStr), zap.Error(err))
+		return nil, err
+	}
+
+	n := big.Int(*m.Number)
+	return &common.NewBlock{
+		Number: &n,
+		Hash:   m.Hash,
+	}, nil
+}
+
```

Comment:
Might be nice to add a comment somewhere about why `isBlockFinalized` is even necessary. I know why now, but someone new might not understand why.